File: September/14 09 25/permutations2.py
#O(n!)
from typing import List

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()  # sort to handle duplicates
        
        def backtrack(path, used):
            if len(path) == len(nums):
                res.append(path[:])
                return
            for i in range(len(nums)):
                if used[i]:
                    continue
                if i > 0 and nums[i] == nums[i-1] and not used[i-1]:
                    continue
                used[i] = True
                path.append(nums[i])
                backtrack(path, used)
                path.pop()
                used[i] = False
        
        backtrack([], [False]*len(nums))
        return res

sol = Solution()
print(sol.permuteUnique([1,1,2]))
# [[1,1,2],[1,2,1],[2,1,1]]

File: September/14/permutations2.swift
class Solution {
    func permuteUnique(_ nums: [Int]) -> [[Int]] {
        var res = [[Int]]()
        let nums = nums.sorted()
        var used = [Bool](repeating: false, count: nums.count)
        
        func backtrack(_ path: [Int]) {
            if path.count == nums.count {
                res.append(path)
                return
            }
            for i in 0..<nums.count {
                if used[i] { continue }
                if i > 0 && nums[i] == nums[i-1] && !used[i-1] { continue }
                used[i] = true
                var newPath = path
                newPath.append(nums[i])
                backtrack(newPath)
                used[i] = false
            }
        }
        
        backtrack([])
        return res
    }
}

let sol = Solution()
print(sol.permuteUnique([1,1,2]))
// [[1,1,2],[1,2,1],[2,1,1]]
